src/Rule/AbstractCompare.php

```diff
@@ -0,0 +1,277 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Validator\Rule;
+
+use Closure;
+use InvalidArgumentException;
+use Yiisoft\Validator\Rule\Trait\SkipOnEmptyTrait;
+use Yiisoft\Validator\Rule\Trait\SkipOnErrorTrait;
+use Yiisoft\Validator\Rule\Trait\WhenTrait;
+use Yiisoft\Validator\RuleWithOptionsInterface;
+use Yiisoft\Validator\SkipOnEmptyInterface;
+use Yiisoft\Validator\SkipOnErrorInterface;
+use Yiisoft\Validator\WhenInterface;
+
+/**
+ * Abstract base for all the comparison validation rules.
+ *
+ * The value being compared with {@see AbstractCompare::$targetValue} or {@see AbstractCompare::$targetAttribute}, which
+ * is set in the constructor.
+ *
+ * It supports different comparison operators, specified
+ * via the {@see AbstractCompare::$operator}.
+ *
+ * The default comparison is based on string values, which means the values
+ * are compared byte by byte. When comparing numbers, make sure to change {@see AbstractCompare::$type} to
+ * {@see AbstractCompare::TYPE_NUMBER} to enable numeric comparison.
+ *
+ * @see CompareHandler
+ *
+ * @psalm-import-type WhenType from WhenInterface
+ */
+abstract class AbstractCompare implements
```

Comment:
Makes using constants a bit uglier:

```php
Compare::TYPE_STRING;
AbstractCompare::TYPE_STRING;
```

Reply:
But that's probably OK.

Reply:
We can create simple class:

```php
final class CompareType {
    const STRING = 'string';
    const NUMBER = 'number';
}
```

This is resolve warning in IDE and make clearly appearance: 

```php
CompareType::STRING
```